File: server/src/main/java/com/github/bin/util/DiceResult.java
package com.github.bin.util;

import lombok.Getter;
import lombok.val;

import java.util.Arrays;
import java.util.Random;

/**
 * @author bin
 * @since 2023/08/22
 */
@Getter
public class DiceResult {
    private final int size;
    private final int max;
    private long sum;
    private final int[] list;

    public DiceResult(int size, int max) {
        this.size = Math.max(Math.min(size, 999), 1);
        this.max = Math.max(Math.min(max, 999_999_999), 1);
        this.sum = this.size;
        this.list = new int[this.size];
        Arrays.fill(this.list, 1);
    }

    public DiceResult(long sum, int[] list, int max) {
        this.size = list.length;
        this.max = Math.max(max, 1);
        this.sum = sum;
        this.list = list;
    }

    public String getOrigin() {
        return String.format("%dd%d", this.size, this.max);
    }

    public void dice() {
        long sum = 0;
        val random = new Random();
        val max = this.max + 1;
        for (int i = 0; i < list.length; i++) {
            val it = random.nextInt(1, max);
            list[i] = it;
            sum += it;
        }
        this.sum = sum;
    }

    public DiceResult plus(DiceResult dice) {
        if (this.max != dice.max) {
            throw new IllegalArgumentException("max must be equal");
        }
        int max = this.max;
        int[] list = new int[this.size + dice.size];
        System.arraycopy(this.list, 0, list, 0, this.size);
        System.arraycopy(dice.list, 0, list, this.size, dice.size);
        return new DiceResult(sum + dice.sum, list, max);
    }

    @Override
    public String toString() {
        return "<" + getOrigin() + ">";
    }
}

File: src/main/kotlin/com/github/bin/util/DiceResult.kt
package com.github.bin.util

class DiceResult(
	/**总数*/
	size: Int,
	/**最大值*/
	max: Int,
) {

	@JvmField
	val max = max.coerceAtLeast(1)

	@JvmField
	val size: Int = when {
		size < 1 -> 1
		size > 999 -> 999
		else -> size
	}

	@JvmField
	var sum: Long = this.size.toLong()

	@JvmField
	val list = IntArray(this.size) { 1 }

	constructor(
		sum: Long,
		/**列表*/
		list: IntArray,
		max: Int,
	) : this(list.size, max) {
		this.sum = sum
		System.arraycopy(list, 0, this.list, 0, size)
	}

	val origin get() = "${list.size}d${max}"

	fun dice(): DiceResult {
		val range = 1..when {
			max < 1 -> return this
			max > 999_999_999 -> 999_999_999
			else -> max
		}
		var sum: Long = 0
		for (i in list.indices) {
			val it = range.random()
			list[i] = it
			sum += it
		}
		this.sum = sum
		return this
	}

	operator fun plus(dice: DiceResult): DiceResult {
		return DiceResult(
			sum = sum + dice.sum,
			max = if (max == dice.max) max else error("max 需要相同"),
			list = intArrayOf(*list, *dice.list)
		)
	}

	override fun toString(): String = "<$origin>"
}
